netket/experimental/operator/_fermions_2nd.py

```diff
@@ -0,0 +1,769 @@
+            raise ValueError(
+                f"Can only add identical hilbert spaces (got A+B, A={self.hilbert}, B={other.hilbert})"
+            )
+        if not np.can_cast(_dtype(other), self.dtype, casting="same_kind"):
+            raise ValueError(
+                f"Cannot add inplace operator with dtype {type(other)} "
+                f"to operator with dtype {self.dtype}"
+            )
+        for t, w in other._operators.items():
+            if t in self._operators.keys():
+                self._operators[t] += w
+            else:
+                self._operators[t] = w
+        self._constant += other._constant
+        return self
+
+    def __sub__(self, other):
+        return self + (-other)
+
+    def __rsub__(self, other):
+        return other + (-self)
+
+    def __neg__(self):
+        return -1 * self
+
+    def __rmul__(self, scalar):
+        return self * scalar
+
+    def __imul__(self, scalar):
+        if not isinstance(scalar, numbers.Number):
+            # we will overload this as matrix multiplication
+            self._iop__matmul__(scalar)
+        if not np.can_cast(_dtype(scalar), self.dtype, casting="same_kind"):
+            raise ValueError(
+                f"Cannot multiply inplace scalar with dtype {type(scalar)} "
+                f"to operator with dtype {self.dtype}"
+            )
+        scalar = np.array(scalar, dtype=self.dtype).item()
+        self._operators = tree_map(lambda x: x * scalar, self._operators)
+        self._constant *= scalar
+        return self
+
+    def __mul__(self, scalar):
+        if not isinstance(scalar, numbers.Number):
+            # we will overload this as matrix multiplication
+            return self._op__matmul__(scalar)
+        dtype = np.promote_types(self.dtype, _dtype(scalar))
+        op = self.copy(dtype=dtype)
+        op *= scalar
+        return op
+
+
+def _convert_terms_to_spin_blocks(terms, n_orbitals, n_spin_components):
+    """see explanation in from_openfermion in conversion between conventions of netket and openfermion"""
+
+    if n_spin_components == 1:
+        return terms
+
+    def _convert_loc(of_loc):
+        orb_idx = of_loc // n_spin_components
+        spin_idx = of_loc % n_spin_components
+        return orb_idx + n_orbitals * spin_idx
+
+    def _convert_term(term):
+        return tuple([(_convert_loc(t[0]), t[1]) for t in term])
+
+    return tuple(list(map(_convert_term, terms)))
+
+
+def _collect_constants(terms, weights):
+    """
+    Openfermion has the convention to store constants as empty terms
+    Returns new terms and weights list, and the collected constants
+    """
+    new_terms = []
+    new_weights = []
+    constant = 0.0
+    for t, w in zip(terms, weights):
+        if len(t) == 0:
+            constant += w
+        else:
+            new_terms.append(t)
+            new_weights.append(w)
+    return new_terms, new_weights, constant
+
+
+def _canonicalize_input(terms, weights, constant, dtype):
```

Comment:
maybe for the future: it might make sense to store all operators in normal-ordering so that it's easier to detect duplciates?

And then it's easier to do this in `_canonicalize_input`. So we never check it again after the operator is created.

Reply:
Yeah, we could do this. Then we would have to store a _is_normal_ordered so we never reorder again. However, I think it's better to leave this normal_ordering purely for the is_hermitian, since that's only called once. We can mostly assume that there are no duplicates, I'd think..

Reply:
ok, let's not do this for now...